cart/redis/infrastructure/redisStorage.go

```diff
@@ -0,0 +1,203 @@
+	},
+) *RedisStorage {
+	r.serializer = serializer
+
+	if config != nil {
+		var err error
+
+		r.ttlGuest, err = time.ParseDuration(config.RedisTTLGuest)
+		if err != nil {
+			panic("can't parse commerce.cart.redis.ttl.guest")
+		}
+
+		r.ttlCustomer, err = time.ParseDuration(config.RedisTTLCustomer)
+		if err != nil {
+			panic("can't parse commerce.cart.redis.ttl.customer")
+		}
+
+		r.keyPrefix = config.RedisKeyPrefix
+
+		var tlsConfig *tls.Config
+		if config.RedisTLS {
+			tlsConfig = &tls.Config{MinVersion: tls.VersionTLS12}
+		}
+
+		r.client = redis.NewClient(&redis.Options{
+			Network:   config.RedisNetwork,
+			Addr:      config.RedisAddress,
+			Password:  config.RedisPassword,
+			DB:        config.RedisDatabase,
+			PoolSize:  int(config.RedisIdleConnections),
+			TLSConfig: tlsConfig,
+		})
+
+		// close redis client
+		runtime.SetFinalizer(r, func(r *RedisStorage) { _ = r.client.Close() })
+	}
+
+	return r
+}
+
+// GetCart fetches a cart from redis and deserializes it
+func (r *RedisStorage) GetCart(ctx context.Context, id string) (*cartDomain.Cart, error) {
+	cmd := r.client.Get(ctx, r.keyPrefix+id)
+	if cmd.Err() != nil {
+		return nil, fmt.Errorf("could not get cart: %w", cmd.Err())
+	}
+
+	b, err := cmd.Bytes()
+	if err != nil {
+		return nil, fmt.Errorf("could not get cart: %w", err)
+	}
+
+	cart, err := r.serializer.Deserialize(b)
+	if err != nil {
+		return nil, fmt.Errorf("could not get cart: %w", err)
+	}
+
+	return cart, nil
+}
+
+// HasCart checks if the cart id exists as a key in redis
+func (r *RedisStorage) HasCart(ctx context.Context, id string) bool {
+	return r.client.Exists(ctx, r.keyPrefix+id).Val() > 0
+}
+
+// StoreCart serializes a cart and stores it in redis
+func (r *RedisStorage) StoreCart(ctx context.Context, cart *cartDomain.Cart) error {
+	if cart == nil {
+		return ErrCartIsNil
+	}
+
+	b, err := r.serializer.Serialize(cart)
+	if err != nil {
+		return fmt.Errorf("could not store cart: %w", err)
+	}
+
+	err = r.client.Set(ctx, r.keyPrefix+cart.ID, b, r.ttl(cart)).Err()
+	if err != nil {
+		return fmt.Errorf("could not store cart: %w", err)
+	}
+
+	return nil
+}
+
+// ttl may differ for guest and customer carts
+func (r *RedisStorage) ttl(cart *cartDomain.Cart) time.Duration {
+	if cart.BelongsToAuthenticatedUser {
+		return r.ttlCustomer
+	}
+
+	return r.ttlGuest
+}
+
+// RemoveCart deletes a cart from redis
+func (r *RedisStorage) RemoveCart(ctx context.Context, cart *cartDomain.Cart) error {
+	if cart == nil {
+		return ErrCartIsNil
+	}
+
+	err := r.client.Del(ctx, r.keyPrefix+cart.ID).Err()
+	if err != nil {
+		return fmt.Errorf("could not remove cart: %w", err)
+	}
+
+	return nil
+}
+
+// Status healthcheck via ping
+func (r *RedisStorage) Status() (alive bool, details string) {
+	err := r.client.Ping(context.Background()).Err()
+	if err != nil {
+		return false, fmt.Errorf("error trying to PING: %w", err).Error()
```

Comment:
maybe mention redis here as well so we know what tried to ping

e.g. redis cart storage ping failed: